runtime/lua/vim/lsp/semantic_tokens.lua

```diff
@@ -94,15 +95,34 @@ end
 ---
 ---@private
 ---@return STTokenRange[]
-local function tokens_to_ranges(data, bufnr, client)
+local function tokens_to_ranges(data, bufnr, client, request)
   local legend = client.server_capabilities.semanticTokensProvider.legend
   local token_types = legend.tokenTypes
   local token_modifiers = legend.tokenModifiers
   local ranges = {}
 
+  local start = uv.hrtime()
+  local ns_to_ms = 0.000001
+  local co = coroutine.running()
+
   local line
   local start_char = 0
   for i = 1, #data, 5 do
+    local elapsed_ms = (uv.hrtime() - start) * ns_to_ms
```

Comment:
I wonder, how many times can this line be executed? 100 times? 1k, 10k, 100k?

Reply:
Can it? Or will it?

It'll run once per token per request (unless it quits early). `hrtime()` uses a vDSO function (`clock_gettime()`), so is pretty cheap without much overhead (no syscall/kernel context switch involved). And it's pretty self regulating--if there ends up being some overhead to this call, the loop will just yield more often but input latency should remain undegraded.

Reply:
What I'm asking I guess is on how many iterations of this loop you start noticing the lag, just to get a rough idea on what's happening

Reply:
It depends on how long the loop takes to run. This is highly dependent on your machine and the types of tokens that are returned. Turning the bitfield for modifiers into the list of modifier names is not a constant operation so each token takes a variable amount of instructions.

That said, here are some stats I gathered on my M1 Pro Macbook Pro. All stats are taken with a release build, using the rust file from the issue I linked in the description.

Without the coroutine/yielding, here's how long it takes my machine to simply process the whole token array into a highlight range list:
```
93.688166 ms to process 17602 tokens (0.005323 ms/token)
116.935833 ms to process 17603 tokens (0.006643 ms/token)
108.793209 ms to process 17604 tokens (0.006180 ms/token)
117.514709 ms to process 17604 tokens (0.006675 ms/token)
116.531417 ms to process 17604 tokens (0.006620 ms/token)
110.875750 ms to process 17605 tokens (0.006298 ms/token)
113.007000 ms to process 17605 tokens (0.006419 ms/token)
```
I was just typing a few characters to feel the lag and get additional token responses sent. Each line above was a full token response from the server.

With that same file, I re-enabled the yield-every-5-ms logic and recorded the number of iterations it completed per wakeup:

```
completed 305 tokens for this loop iteration
completed 465 tokens for this loop iteration
completed 1045 tokens for this loop iteration
completed 1091 tokens for this loop iteration
completed 1100 tokens for this loop iteration
completed 997 tokens for this loop iteration
completed 880 tokens for this loop iteration
completed 675 tokens for this loop iteration
completed 773 tokens for this loop iteration
completed 1081 tokens for this loop iteration
completed 1081 tokens for this loop iteration
completed 1085 tokens for this loop iteration
completed 1054 tokens for this loop iteration
completed 872 tokens for this loop iteration
completed 890 tokens for this loop iteration
completed 790 tokens for this loop iteration
completed 453 tokens for this loop iteration
completed 958 tokens for this loop iteration
completed 1080 tokens for this loop iteration
19 yields to process response
```
This was for _one_ response from the server.

I think the sensitivity to input lag varies from person to person, but for the sake of demonstration, lets say we shoot for 60fps, which most people can agree is very responsive. This means that only ~16.7ms can be spent doing work between handling an input key and drawing it to the screen. That's for _everything_ it has to do (since we do all our work on the main thread).

Yielding every 5 ms seems like a good compromise to leave a lot of headroom for neovim to do other stuff per loop iteration. We could go higher, but we start to crowd out the available time for that other stuff before the input latency becomes noticeable.

Reply:
It's also important to note that this PR enables us to stop processing stale tokens whereas before we always processed everything. So whilst being more responsive, it's also a lot more efficient.

Reply:
To be fair, HEAD is currently synchronous once the response arrives, and we only process the response if it isn't stale.

But since this PR makes things makes more responsive, we won't be processing as many tokens during high edit activity--so in that sense, it's much more efficient.

Reply:
> To be fair, HEAD is currently synchronous once the response arrives, and we only process the response if it isn't stale.

But since we're now scheduling over several event loop iterations, isn't it possible for the processing of tokens from multiple responses to interleave? (Assuming we don't abandon). Because we can abandon, we can now end up only partially processing responses?

Reply:
They can't interleave. As soon as a new request is sent (due to the buffer changing and the debounce timer expiring), the `active_request` field is cleared which effectively kills the current coroutine that's processing its response.

So yeah, we will be abandoning partial responses if there's heavy edit activity. This PR makes it a lot closer to asynchronous!

Reply:
Yes, I did qualify with: "assuming we don't abandon".

Anyway, good work. I plan on doing something very similar with treesitter.

Reply:
Cool, thanks!